module/input/Microphone/src/Microphone.cpp

```diff
@@ -0,0 +1,367 @@
+bool voice2json_transcribe_wav(MicProcHandles *handles) {
+    char *const args[] = {
+        (char*) "python",     
+        (char*) "-m",
+        (char*) "voice2json",
+        (char*) "--base-directory",
+        (char*) "/home/nubots/voice2json/",
+        (char*) "-p",
+        (char*) "en",
+        (char*) "transcribe-wav",
+        (char*) "--input-size",
+        NULL
+    };
+    
+    return spawn_process(handles, args);
+}
+
+char *voice2json_recognize_intent(char *input) {
+    char *output = 0;
+    char buffer[0x1000];
+    ssize_t bytes_read = 0;
+    
+    char *const args[] = {
+        (char*) "python",     
+        (char*) "-m",
+        (char*) "voice2json",
+        (char*) "--base-directory",
+        (char*) "/home/nubots/voice2json/",
+        (char*) "-p",
+        (char*) "en",
+        (char*) "recognize-intent",
+        NULL
+    };
+    
+    MicProcHandles handles = {};
+    if(spawn_process(&handles, args)) {
+        write(handles.stdin, input, strlen(input));
+        
+        bytes_read = read(handles.stdout, buffer, sizeof(buffer)-1);
+        assert(bytes_read > 0);
+        buffer[bytes_read] = 0;
+        
+        output = strdup(buffer);
+    }
+    if(handles.stdout) close(handles.stdout);
+    if(handles.stderr) close(handles.stderr);
+    if(handles.stdin) close(handles.stdin);
+
+    return output;
+}
+
+void write_audio_to_file(int fd, char *filename) {
+    FILE *file = fopen(filename, "rb");
+    fseek(file, 0, SEEK_END);
+    int file_size = ftell(file);
+    fseek(file, 0, SEEK_SET);
+    assert(file_size != -1);
+    
+    char *file_buffer = (char*) malloc(file_size);
+    fread(file_buffer, file_size, 1, file);
+    fclose(file);
+    
+    char file_size_buf[64];
+    sprintf(file_size_buf, "%d\n", file_size);
+        
+    write(fd, file_size_buf, strlen(file_size_buf));
+    
+    ssize_t write_res = write(fd, file_buffer, file_size);
+    assert(write_res == file_size);
+
+    free(file_buffer);
+}
+
+
+Microphone::Microphone(std::unique_ptr<NUClear::Environment> environment) : Reactor(std::move(environment)), config{} {
+    
+    on<Configuration>("Microphone.yaml").then([this](const Configuration& cfg) {
+        // Use configuration here from file Microphone.yaml
+        this->log_level = cfg["log_level"].as<NUClear::LogLevel>();
+    });
+    
+
+    
+    if(!voice2json_transcribe_wav(&handles)) {
+        perror("failed to start voice2json");
+        return;
+    }
+    
+    on<Trigger<MicControlMsg>>().then([this](const MicControlMsg& msg) {
+        switch(msg.type) {
+            case MIC_MSG_ENABLE: 
+                enabled = true;
+            break;
+            case MIC_MSG_DISABLE:
+                enabled = false;
+            break;
+            case MIC_MSG_TEST_AUDIO:
+                char filename[1024];
+                sprintf(filename, "/home/nubots/nuhear/audio/voice_%03d.wav", msg.data);
```

Comment:
Could we perhaps enter the full path to a file instead of an integer for the number? I'd like to be able to test it with files located somewhere else on the file system.

Reply:
Changed to accept file paths now. Can be tested by typing "file /home/nubots/nuhear/audio/voice_001.wav" for example.